File: kotlin/binary-search/main_2805.kt
import kotlin.math.max

fun main() {
    val (_, length) = readLine()!!.split(' ').map { it.toLong() }
    val tree = readLine()!!.split(' ').map{ it.toInt() }

    var start = 1L
    var end = tree.maxOrNull()?.toLong() ?: 0L
    var ans = 0L
    while (start <= end) {
        val mid = (start + end) / 2
        if (mid == 0L) break
        if (tree.sumOf { if (it - mid > 0) it - mid else 0 } >= length) {
            ans = max(ans, mid)
            start = mid + 1
        } else {
            end = mid - 1
        }
    }
    print(ans)
}

File: cpp/binary-search/main_2805.cpp
#include <iostream>
#include <vector>
using namespace std;

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(nullptr);
	cout.tie(nullptr);

	int count, limit;
	cin >> count >> limit;
	vector<int> input(count);
	long long mx = 0;
	for (int i = 0; i < count; ++i) {
		cin >> input[i];
		mx = max(mx, static_cast<long long>(input[i]));
	}

	long long start = 1;
	long long end = mx;
	long long ans = 0;
	while (start <= end) {
		const long long mid = (start + end) / 2;
		long long sum = 0;
		for (auto value : input) {
			sum += max(value - mid, 0LL);
			if (sum >= limit) {
				break;
			}
		}
		if (sum < limit) {
			end = mid - 1;
		} else {
			ans = mid;
			start = mid + 1;
		}
	}
	cout << ans;
}
